File: CPU_SCHEDULING_ALGORITHMS/RR.cpp
// Hello ashlok keep going....
#include <iostream>
using namespace std;

struct Process
{
    int id;
    int burstTime;
    int arrivalTime;
    int completionTime;
    int turnaroundTime;
    int waitingTime;
};

void calculateTimes(Process processes[], int n, int timeQuantum)
{
    int remainingTime[n];
    // Storing the original burst time for final usage
    for (int i = 0; i < n; ++i)
        remainingTime[i] = processes[i].burstTime;

    int currentTime = 0;
    bool isCompleted = false; // while all processes are completed tracking variable

    while (!isCompleted)
    {
        isCompleted = true;
        for (int i = 0; i < n; ++i)
        {
            if (remainingTime[i] > 0)
            {
                isCompleted = false;
                if (remainingTime[i] > timeQuantum)
                {
                    currentTime += timeQuantum;
                    remainingTime[i] -= timeQuantum;
                }
                else
                {
                    currentTime += remainingTime[i];
                    remainingTime[i] = 0;
                    processes[i].completionTime = currentTime;
                }
            }
        }
    }
}

void calculateTurnAroundTime(Process processes[], int n)
{
    for (int i = 0; i < n; ++i)
        processes[i].turnaroundTime = processes[i].completionTime - processes[i].arrivalTime;
}

void calculateWaitingTime(Process processes[], int n)
{
    for (int i = 0; i < n; ++i)
        processes[i].waitingTime = processes[i].turnaroundTime - processes[i].burstTime;
}

void printTable(Process processes[], int n)
{
    cout << "Processes: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].id << " ";
    cout << '\n';
    cout << "Arrival Time: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].arrivalTime << " ";
    cout << '\n';
    cout << "Burst Time: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].burstTime << " ";
    cout << '\n';
    cout << "Completion Time: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].completionTime << " ";
    cout << '\n';
    cout << "Turn Around Time: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].turnaroundTime << " ";
    cout << '\n';
    cout << "Waiting Time: ";
    for (int i = 0; i < n; ++i)
        cout << processes[i].waitingTime << " ";
    cout << '\n';
}

int main()
{
    int n, timeQuantum;
    cout << "Enter No. of Processes and Time Quantum: ";
    cin >> n >> timeQuantum;

    Process processes[n];
    for (int i = 0; i < n; ++i)
    {
        cout << "Process " << (i + 1) << " \n";
        processes[i].id = i + 1;
        cout << "Enter Arrival Time: ";
        cin >> processes[i].arrivalTime;
        cout << "Enter Burst Time: ";
        cin >> processes[i].burstTime;
    }

    calculateTimes(processes, n, timeQuantum);
    calculateTurnAroundTime(processes, n);
    calculateWaitingTime(processes, n);

    cout << "Round Robin Results: \n";
    printTable(processes, n);
    return 0;
}

/*
1. Implementing FCFS scheduling algorithm in Java
2. Implementing SJF (with no premtion) scheduling alogithm
3. Implementing RR Scheduling Algorithm in Java
4. Write a Java program that implements banker's alogithm
5. Write a Java program that impolements FIFO page replacement algorithm
6. Write a Java Program that implements LRU Page replacement algorithm
7. Design a File System in Java
8. Process Commmunication
    1. Implement the producer - consumer problem using shared memory
    2. Implement the producer - consumer problem using message passing
    3. One form of communiaction in a client - server systems environment in Remote Method invocation
9. Threads
    1.
    2.
10. Synchronization
    1. Implementing Bounded buffer problem in Java
    2. Implementing the reader - writers using Java synchronisation
    3. The sleeping barber problem
*/

File: CPU_SCHEDULING_ALGORITHMS/RR.java
package Practice;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class RR {
    public static class Process {
        int process;
        int burst;
        int arrival;
        int complete;
        int waiting;
        int turnaroundTime;

        public Process(int process, int arrival, int burst) {
            this.process = process;
            this.burst = burst;
            this.arrival = arrival;
            this.complete = 0;
            this.waiting = 0;
            this.turnaroundTime = 0;
        }

        @Override
        public String toString() {
            return "Process: " + process + " Arrival: " + arrival + " Burst: " + burst + " Completion Time: " + complete
                    +
                    " Waiting: " + waiting + " Turn Around Time: " + turnaroundTime;
        }
    }

    public static void run(List<Process> x, int timeQuantum) {
        int n = x.size();
        int[] remainingTime = new int[n];

        for (int i = 0; i < n; ++i) {
            remainingTime[i] = x.get(i).burst;
        }

        int currentTime = 0;

        while (true) {
            boolean allCompleted = true;

            for (int i = 0; i < n; ++i) {
                if (remainingTime[i] > 0) {
                    allCompleted = false;

                    if (remainingTime[i] > timeQuantum) {
                        currentTime += timeQuantum;
                        remainingTime[i] -= timeQuantum;
                    } else {
                        currentTime += remainingTime[i];
                        remainingTime[i] = 0;
                        x.get(i).complete = currentTime;
                    }
                }
            }

            if (allCompleted) {
                break;
            }
        }

        for (Process a : x) {
            a.turnaroundTime = a.complete - a.arrival;
            a.waiting = a.turnaroundTime - a.burst;
        }
    }

    public static void main(String[] args) {
        List<Process> x = new ArrayList<>();
        x.add(new Process(1, 0, 8));
        x.add(new Process(2, 5, 2));
        x.add(new Process(3, 1, 7));
        x.add(new Process(4, 6, 3));
        x.add(new Process(5, 8, 5));

        Collections.sort(x, (a, b) -> a.arrival - b.arrival);

        run(x, 3);

        for (Process a : x) {
            System.out.println(a.toString());
        }
    }
}
